### PR TITLE
[RFC] Preserve empty lines between array elements

### PR DESCRIPTION
**Pull request by vjeux**
This has come up many times in the past and while going through the fb codebase, there are a few instances where we group array elements logically using empty lines and it's a shame that they are gone.

### CODE REVIEW

**Comment by garth** [comment]
in cerebraljs (http://www.cerebraljs.com/) arrays and object literals are used to compose a tree of actions, which read like high level recipes of what happens in the app.

Being able to place the each item on it's own line is key to readability.

I'm sure that there are many use cases where an option to "preserve-new-lines" (even on a per file bases) would help to maintain readability.

**Comment by vjeux** [comment]
@garth could you open an issue with real cerebral code where it outputs things that you are not happy with? This way we have concrete examples to look at in order to find a way to print it the best way :)

File: src/printer.js [hunk 1]
```diff
@@ -764,16 +764,29 @@ function genericPrintNoParens(path, options, print) {
...
+        let separatorParts = [];
+        path.each(
+          function(childPath) {
+            printedElements.push(concat(separatorParts));
+            printedElements.push(group(print(childPath)));
+
+            separatorParts = [",", line];
+            if (childPath.getValue() &&
+                util.isNextLineEmpty(options.originalText, childPath.getValue())) {
+              separatorParts.push(hardline);
```
Comment (jlongster) [inline src/printer.js]: This also effectively forces arrays to be expanded if they were expanded in the original source.

**Comment by jlongster** [comment]
Did you mean to use `hardline` instead of `softline`, where it would maintain existing line breaks *if* it stayed expanded? Or did you intentionally want to keep arrays expanded?

I feared this was going to come up as arrays are inconsistent with objects in this regard. I personally wish we wouldn't do this, and I feel the same about objects.

But it's only my personal preference and I don't have a strong technical reason for it. It makes it so that if you add something, and it makes it expand, and then you remove it, it doesn't collapse again. I've definitely run into this already with objects where I had to manually collapse them.

I'm fine adapting to the community thought and doing what's needed. It's not a huge deal.

**Comment by vjeux** [comment]
Oops, I just wanted to preserve new lines in between, not force it to be expanded if there was a single newline. I copy and pasted the object code.

**Comment by vjeux** [comment]
Just replaced `hardline` with `softline` and added a test to make sure that if the array was not expanded, those empty lines vanish.

**Comment by jlongster** [comment]
Ok, cool thanks!

**Pull request merged**